services/QuillLMS/app/controllers/auth/google_controller.rb

```diff
@@ -1,28 +1,43 @@
 class Auth::GoogleController < ApplicationController
-  before_action :set_profile,                     only: :google
-  before_action :set_user,                        only: :google
-  before_action :save_teacher_from_google_signup, only: :google
-  before_action :save_student_from_google_signup, only: :google
-  before_action :follow_google_redirect,          only: :google
-
-  def google
-    if @user.teacher?
-      GoogleStudentImporterWorker.perform_async(@user.id, 'Auth::GoogleController')
-    end
+  before_action :set_profile, only: [:authorized_and_authenticated, :authenticated]
+  before_action :verify_credentials, only: :authenticated
+  before_action :set_user, only: [:authorized_and_authenticated, :authenticated]
+  before_action :save_teacher_from_google_signup, only: [:authorized_and_authenticated, :authenticated]
+  before_action :save_student_from_google_signup, only: [:authorized_and_authenticated, :authenticated]
+  before_action :follow_google_redirect, only: [:authorized_and_authenticated, :authenticated]
 
-    if @user.student?
-      GoogleStudentClassroomWorker.perform_async(@user.id)
-    end
+  def authorized_and_authenticated
+    run_background_jobs
+    sign_in(@user)
+    redirect_to_profile_or_post_auth
+  end
 
+  def authenticated
+    run_background_jobs
     sign_in(@user)
+    redirect_to_profile_or_post_auth
+  end
 
+  private def redirect_to_profile_or_post_auth
     if session[ApplicationController::POST_AUTH_REDIRECT].present?
       url = session[ApplicationController::POST_AUTH_REDIRECT]
       session.delete(ApplicationController::POST_AUTH_REDIRECT)
-      return redirect_to url
+      redirect_to url
+    else
+      redirect_to profile_path
     end
+  end
 
-    redirect_to profile_path
+  private def verify_credentials
+    user = User.where('google_id = ? OR email = ?', @profile.google_id&.to_s, @profile.email&.downcase).first
+    return if user.nil?
```

Comment:
Do we need this return if we have the redirect on line 35?

Reply:
Since the redirect is an `unless` we actually want to exit at this point.